src/solc_0.7/GameToken/GameToken.sol

```diff
@@ -0,0 +1,534 @@
+        uint256 baseId,
+        uint256[] memory assetIds,
+        uint256[] memory amounts
+    ) internal {
+        if (assetIds.length == 0) {
+            return;
+        }
+        require(assetIds.length == amounts.length, "INVALID_INPUT_LENGTHS");
+        uint256 currentValue;
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            currentValue = _gameAssets[baseId][assetIds[i]];
+            require(amounts[i] != 0, "INVALID_ASSET_ADDITION");
+            _gameAssets[baseId][assetIds[i]] = currentValue.add(amounts[i]);
+        }
+        if (assetIds.length == 1) {
+            _asset.safeTransferFrom(from, address(this), assetIds[0], amounts[0], "");
+        } else {
+            _asset.safeBatchTransferFrom(from, address(this), assetIds, amounts, "");
+        }
+    }
+
+    /// @notice Remove assets from a GAME.
+    /// @param baseId The baseId of the GAME to remove assets from.
+    /// @param assetIds An array of asset Ids to remove.
+    /// @param values An array of the number of each asset id to remove.
+    /// @param to The address to send removed assets to.
+    function _removeAssets(
+        uint256 baseId,
+        uint256[] memory assetIds,
+        uint256[] memory values,
+        address to
+    ) internal {
+        if (assetIds.length == 0) {
+            return;
+        }
+        require(assetIds.length == values.length && assetIds.length != 0, "INVALID_INPUT_LENGTHS");
+        uint256 currentValue;
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            currentValue = _gameAssets[baseId][assetIds[i]];
+            require(currentValue != 0 && values[i] != 0 && values[i] <= currentValue, "INVALID_ASSET_REMOVAL");
+            _gameAssets[baseId][assetIds[i]] = currentValue.sub(values[i]);
+        }
+
+        if (assetIds.length == 1) {
+            _asset.safeTransferFrom(address(this), to, assetIds[0], values[0], "");
+        } else {
+            _asset.safeBatchTransferFrom(address(this), to, assetIds, values, "");
+        }
+    }
+
+    /// @notice Get the creator of the token type `id`.
+    /// @param id The id of the token to get the creator of.
+    /// @return the creator of the token type `id`.
+    function creatorOf(uint256 id) public view override returns (address) {
+        require(id != uint256(0), "GAME_NEVER_MINTED");
+        address originalCreator = address(id / CREATOR_OFFSET_MULTIPLIER);
+        address newCreator = _creatorship[originalCreator];
+        if (newCreator != address(0)) {
+            return newCreator;
+        }
+        return originalCreator;
+    }
+
+    /// @notice Return the URI of a specific token.
+    /// @param gameId The id of the token.
+    /// @return uri The URI hash of the token.
+    function tokenURI(uint256 gameId) public view override returns (bytes32 uri) {
+        require(_ownerOf(gameId) != address(0), "BURNED_OR_NEVER_MINTED");
+        uint256 baseId = _storageId(gameId);
+        return _metaData[baseId];
```

Comment:
This need to return a uri.
to do that we can reuse ERC1155ERC721.sol implementation of ipfs uri 
see : https://github.com/thesandboxgame/sandbox-smart-contracts/blob/0e8428debf30cec3f2f575da9d8255daf71992f9/src/solc_0.5/Asset/ERC1155ERC721Test.sol#L1115
and https://github.com/thesandboxgame/sandbox-smart-contracts/blob/0e8428debf30cec3f2f575da9d8255daf71992f9/src/solc_0.5/Asset/ERC1155ERC721Test.sol#L1075

Reply:
@wighawag I would probably just copy-paste `toFullURI` and modify as needed unless you  would rather this contract be imported to the gameToken?
Also, I would make use of the openzeppelin `toString` function instead of using our own version of this,